system/lib/fetch/emscripten_fetch.cpp

```diff
@@ -66,16 +68,65 @@ emscripten_fetch_t *emscripten_fetch(emscripten_fetch_attr_t *fetch_attr, const
 	}
 
 	emscripten_fetch_t *fetch = (emscripten_fetch_t *)malloc(sizeof(emscripten_fetch_t));
+	if (!fetch) return 0;
 	memset(fetch, 0, sizeof(emscripten_fetch_t));
 	fetch->id = globalFetchIdCounter++; // TODO: make this thread-safe!
 	fetch->userData = fetch_attr->userData;
-	fetch->url = strdup(url); // TODO: free
+	fetch->url = strdup(url);
+	if(!fetch->url)
+	{
+		fetch_free(fetch);
+		return 0;
+	}
 	fetch->__attributes = *fetch_attr;
-	fetch->__attributes.destinationPath = fetch->__attributes.destinationPath ? strdup(fetch->__attributes.destinationPath) : 0; // TODO: free
-	fetch->__attributes.userName = fetch->__attributes.userName ? strdup(fetch->__attributes.userName) : 0; // TODO: free
-	fetch->__attributes.password = fetch->__attributes.password ? strdup(fetch->__attributes.password) : 0; // TODO: free
-	fetch->__attributes.requestHeaders = 0;// TODO:strdup(fetch->__attributes.requestHeaders);
-	fetch->__attributes.overriddenMimeType = fetch->__attributes.overriddenMimeType ? strdup(fetch->__attributes.overriddenMimeType) : 0; // TODO: free
+
+#define STRDUP_OR_ABORT(s)				\
+	if (s)								\
+	{									\
+		s = strdup(s);					\
+		if (!s)							\
+		{								\
+			fetch_free(fetch);			\
+			return 0;					\
+		}								\
+	}
+
+	STRDUP_OR_ABORT(fetch->__attributes.destinationPath);
+	STRDUP_OR_ABORT(fetch->__attributes.userName);
+	STRDUP_OR_ABORT(fetch->__attributes.password);
+	if(fetch->__attributes.requestHeaders)
+	{
+		size_t headersCount;
```

Comment:
This should be `size_t headersCount = 0;` - local variables are not guaranteed to be zero initialized.